### PR TITLE
Feedback

### PR DESCRIPTION
**Pull request by jeajin**
:wave:! GitHub Classroom created this pull request as a place for your teacher to leave feedback on your work. It will update automatically. **Don’t close or merge this pull request**, unless you’re instructed to do so by your teacher.
 In this pull request, your teacher can leave comments and feedback on your code. Click the **Subscribe** button to be notified if that happens.
 Click the **Files changed** or **Commits** tab to see all of the changes pushed to the default branch since the assignment started. Your teacher can see this too.
<details>
<summary>
<strong>Notes for teachers</strong>
</summary>

 Use this PR to leave feedback. Here are some tips:
 - Click the **Files changed** tab to see all of the changes pushed to the default branch since the assignment started. To leave comments on specific lines of code, put your cursor over a line of code and click the blue **+** (plus sign). To learn more about comments, read “[Commenting on a pull request](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/commenting-on-a-pull-request)”.
- Click the **Commits** tab to see the commits pushed to the default branch. Click a commit to see specific changes.
- If you turned on autograding, then click the **Checks** tab to see the results.
- This page is an overview. It shows commits, line comments, and general comments. You can leave a general comment below.
 For more information about this pull request, read “[Leaving assignment feedback in GitHub](https://docs.github.com/education/manage-coursework-with-github-classroom/leave-feedback-with-pull-requests)”.
</details>

Subscribed: @jeajin @82153 @eunseopski @wjddnr0920 @2JAE22 @serin-kimm

### CODE REVIEW

**Review by 2JAE22 — APPROVED** [review]
검토완료. 피드백 부탁드립니다.

File: mmseg_train.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+from mmseg.registry import DATASETS, TRANSFORMS, MODELS, METRICS
+from mmseg.models.segmentors import EncoderDecoderWithoutArgmax
+from mmseg.models.decode_heads import  FCNHeadWithoutAccuracy, SegformerHeadWithoutAccuracy
+from mmseg.models.utils.wrappers import resize
+from mmseg.evaluation.metrics import DiceMetric
+from mmseg.models.losses import LossByFeatMixIn
+from mmengine.runner import Runner, load_checkpoint
+from mmseg.datasets.XRayDataset import CLASSES,IND2CLASS,IMAGE_COUNT_TRAIN
+from mmengine.config import Config, DictAction
+from mmengine.logging import print_log
```
Comment (jsshinn) [inline mmseg_train.py:19]: Runner 를 제외하고 사용되지 않는 package 는 삭제하는게 좋을 것 같습니다.

File: mmseg_test.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+from code.loader.mmseg_config_loader import get_config
+    
+def test(model, data_loader, thr=0.5):
+    model = model.cuda()
+    model.eval()
+
+    rles = []
+    filename_and_class = []
+    with torch.no_grad():
+        n_class = len(CLASSES)
```
Comment (jsshinn) [inline mmseg_test.py:26]: 사용되지 않는 변수입니다

File: mmseg_test.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+    # add output_dir in metric
+    if args.out is not None:
+        cfg.test_evaluator['output_dir'] = args.out
+        cfg.test_evaluator['keep_results'] = True
+
+    # build the runner from config
+    runner = Runner.from_cfg(cfg)
+    model = MODELS.build(cfg.model)
+    checkpoint = load_checkpoint(model, cfg.resume, map_location='cpu')
```
Comment (jsshinn) [inline mmseg_test.py:94]: 사용되지 않는 변수 입니다

File: mmseg_test.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        cfg.tta_model.module = cfg.model
+        cfg.model = cfg.tta_model
+
+    # add output_dir in metric
+    if args.out is not None:
+        cfg.test_evaluator['output_dir'] = args.out
+        cfg.test_evaluator['keep_results'] = True
+
+    # build the runner from config
+    runner = Runner.from_cfg(cfg)
```
Comment (jsshinn) [inline mmseg_test.py:92]: 사용되지 않는 변수입니다

File: streamlit/pages/EDA_viewer.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    st.header("클래스별 포인트 클라우드 시각화")
+    
+    col1, col2 = st.columns(2)
+    
+    with col1:
+        st.subheader("왼손 클래스별 분포")
+        left_clouds = PointCloudGenerator.create_class_point_cloud(left_jsons, sample_shape)
+        for class_name, cloud in left_clouds.items():
+            if cloud.any():  # 비어있지 않은 클래스만 표시
+                st.image(cloud, caption=f"Left - {class_name}", use_column_width=True)
```
Comment (jsshinn) [inline streamlit/pages/EDA_viewer.py:55]: 왼손, 오른손 모두 동일한 코드 인 것 같습니다. 함수화 하여 중복 코드를 줄이면 좋을 것 같습니다.

File: streamlit/main.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
+## main.py ##
+import streamlit as st
+
+# 기본 설정
+st.set_page_config(
+    page_title="Segment_viz 및 Point Cloud 뷰어",
+    layout="wide"
+)
+from pages.EDA_viewer import main as eda_viewer_main
```
Comment (jsshinn) [inline streamlit/main.py:9]: eda_viewer 또한 utils 아래에 다른 viewer 들과 같은 level 로 보이는 것 같은데, utils 로 옮기는게 더 맞지 않을까 싶습니다.

File: streamlit/code/utils/utils.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+    #GPU 에서 결과의 재현성 보장여부
+    torch.backends.cudnn.deterministic = True 
+    
+    #cuDNN최적화된 알고리즘 선택
+    torch.backends.cudnn.benchmark = False
+
+
+
+def set_wandb(configs):
```
Comment (jsshinn) [inline streamlit/code/utils/utils.py:66]: set_wandb 함수가 아래처럼 여러곳에 중복되어있는 것 같습니다! 하나로 정리되는게 좋을 것 같습니다.
- code/utils
- streamlit/code/utils
- UNet++/cv/utils

File: work_dirs/hrnet2/20241125_102705/vis_data/config.py
Comment (jsshinn) [inline work_dirs/hrnet2/20241125_102705/vis_data/config.py:1]: work_dirs/hrnet2/20241125_102847/vis_data/config.py 의 내용과 다를게 없어보입니다! 하나는 삭제하는게 좋겠습니다

File: streamlit/code/loss_functions/loss_selector.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    def __init__(self) -> None:
+        self.loss_classes = {
+            "BCEWithLogitsLoss": CustomBCEWithLogitsLoss,
+            "FocalLoss": FocalLoss,
+            "DiceLoss": DiceLoss,
+            "IOULoss": IOULoss,
+            "CombinedLoss": CombinedLoss,
+        }
+
+    def get_loss(self, loss_name, **loss_parameter):
```
Comment (jsshinn) [inline streamlit/code/loss_functions/loss_selector.py:26]: python 의 [getattr()](https://docs.python.org/ko/3.10/library/functions.html#getattr) 내장 함수를 사용하면 base_loss 의 loss 이름으로 attribute 를 dynamic 하게 가져올 수 있습니다.
```
loss = getattr(base_loss, loss_name)
```

File: code/models/base_model.py
Comment (jsshinn) [inline code/models/base_model.py:1]: 아래 model 들은 모두 forward, load_pretrained_weights, apply_lora 함수를 사용하는 것 같습니다. 하나의 추상화된 class를 만들고, 해당 class를 상속받는게 중복 코드를 막을 수 있을 것 같습니다.

File: code/loader/train_loader.py [hunk 9]
```diff
@@ -0,0 +1,35 @@
+import os
+import albumentations as A
+from torch.utils.data import DataLoader
+from datasets import XRayDataset
+
+def get_data_loaders(config):
+    IMAGE_ROOT_TRAIN = config.data.image_root_train
+    LABEL_ROOT_TRAIN = config.data.label_root_train
+    IMAGE_ROOT_TEST = config.data.image_root_test
```
Comment (jsshinn) [inline code/loader/train_loader.py:9]: 사용되지 않는 것 같습니다!
